bot.py

```diff
@@ -296,6 +304,29 @@ async def create_or_get_voice_client(self, message):
 
         return self.voice_client
 
+    @tasks.loop(seconds=2)
```

Comment:
Can we increase this to at least 1 minute?

Reply:
This could also be done without a loop. You could just schedule a disconnect in the after_callback:

```
async def attempt_disconnect():
     await asyncio.sleep(10min)
     if is_playing():
         # Something is playing meaning another disconnect will be scheduled once that is done.
         return
     disconnect()
     # Or check if enough inactivity has passed. For example by setting self.last_command_time in the after_callback function.
     if something something:
         disconnect()

def after_callback(f):
     ....
     asyncio.create_task(self.attempt_disconnect())
```

Reply:
This is a much better solution, nice!

I was trying to do something similar a while ago but kept having a really hard time accessing the event loop. I came to the conclusion that it was impossible to do so directly because it was being handled by the discord.py library or something. I was just being dumb though and trying to do:
```python
asyncio.create_task()
```
instead of:
```python
self.loop.create_task()
```

The only issue I see with your solution is that if the last `attempt_disconnect()` does it's check before the `DISCONNECT_TIMER` is up, then the bot will potentially never disconnect, due to there being nothing to call a new `attempt_disconnect()`.

However, this can be easily worked around by just not caring about commands and instead only checking for voice activity. And with this solution we can actually do that very easily, so if you guys agree I'll implement this instead.

Reply:
Yes, only implement for voice-related commands.

With the *sleep-then-attempt-disconnect* route in `after_callback`, you only need the sleep timer. If the timer expires while something is playing, it shouldn't disconnect anyway. If the timer expires right after the next song finishes, you could check a (semi-global) token set right before sleeping; if the token is still the same, then the current sleep is the only sleep, and therefore is the *after-queue* sleep.

So something like this:
```python
async def attempt_disconnect():
    self.sleep_token = random.random()
    local_sleep_token = self.sleep_token
    await asyncio.sleep(self.SLEEP_TIMER)
    if self.sleep_token != local_sleep_token:
        # Not the latest disconnect attempt
        return
```

Reply:
Yes that's what I was thinking as well ;)